Fix a typo and address remaining service reference TODO items

dougbu opened the pull request:
- #4923
- typo caused problems when cleaning files
- add `%(OpenApiProjectReference.GlobalPropertiesToRemove)` metadata
- address timing issues cropping up occasionally in builds using service ref features
  - avoid `AfterTargets="Build"`; referencing projects sometimes continue while post-build work is done
    - run after `CoreBuild` in inner builds and after `DispatchToInnerBuilds` in outer builds
    - do same in GetDocumentInsider.csproj
- set only properties in buildMultiTargeting\Microsoft.Extensions.ApiDescription.Server.targets
  - items not evaluated early enough to reference in all cases
- rename Microsoft.Extensions.ApiDescription.Client tasks
- remove net461 task assembly

dougbu committed:
Fix a typo and address remaining service reference TODO items
- #4923
- typo caused problems when cleaning files
- add `%(OpenApiProjectReference.GlobalPropertiesToRemove)` metadata
- address timing issues cropping up occasionally in builds using service ref features
  - avoid `AfterTargets="Build"`; referencing projects sometimes continue while post-build work is done
    - run after `CoreBuild` in inner builds and after `DispatchToInnerBuilds` in outer builds
    - do same in GetDocumentInsider.csproj
- set only properties in buildMultiTargeting\Microsoft.Extensions.ApiDescription.Server.targets
  - items not evaluated early enough to reference in all cases
- rename Microsoft.Extensions.ApiDescription.Client tasks
- remove net461 task assembly

dougbu commented on src/Tools/Extensions.ApiDescription.Client/src/GetCurrentOpenApiReference.cs:
```diff
@@ -9,7 +9,7 @@ namespace Microsoft.Extensions.ApiDescription.Client
     /// Restore <see cref="ITaskItem"/>s from given property value.
     /// </summary>
-    public class GetCurrentItems : Task
+    public class GetCurrentOpenApiReference : Task
```
This one's for you @rynowak 😺

dougbu commented on src/Tools/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.props:
```diff
@@ -98,6 +93,13 @@
+        @(OpenApiProjectReference). These properties, along with Configuration, Platform, RuntimeIdentifier and
+        TargetFrameworks, are also removed when invoking the 'OpenApiGetDocuments' target in the referenced project.
+      -->
+      <GlobalPropertiesToRemove>TargetFramework</GlobalPropertiesToRemove>
```
@rainersigwald is this enough in both contexts where the metadata is used (especially where additional properties aren't always listed)?

rainersigwald replied:
I'm not sure I understand the question, can you elaborate?

The implementation in this PR looks ok to me. Another option would be using the metadata `UndefineProperties`, which is handled internally by the `MSBuild` task and so would come along for free in any invocation. But this is The Standard Pattern.

The only downside I can think of to setting this in the item definition is that it becomes easier to accidentally override. But it also becomes possible to _intentionally_ override, which seems worth it.

dougbu replied:
Your answer covered my questions 😃

dougbu commented on src/Tools/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -25,7 +25,6 @@
-      <GlobalPropertiesToRemove>TargetFramework</GlobalPropertiesToRemove>
       <Private>false</Private>
       <ReferenceOutputAssembly>false</ReferenceOutputAssembly>
       <SkipGetTargetFrameworkProperties>true</SkipGetTargetFrameworkProperties>
```
@rainersigwald should this also set `%(ShouldUnsetParentConfigurationAndPlatform)`?

rainersigwald replied:
Is it expected/allowed to have a different API surface between configurations/platforms? Like say a debug-only endpoint or something?

I would generally _not_ set that; it makes sense to let the user's solution configuration for those things flow here, IMO.

dougbu replied:
Thanks!

dougbu commented on src/Tools/Extensions.ApiDescription.Server/src/build/Microsoft.Extensions.ApiDescription.Server.targets:
```diff
@@ -65,7 +65,7 @@
 
   <Target Name="_GenerateOpenApiDocuments"
-      AfterTargets="Build"
+      AfterTargets="CoreBuild"
```
@rainersigwald is it expected that `<MSBuild/>` tasks calling the `Build` target can return to the project (and continue) before the referenced project does `AfterBuild="Build"` work?

rainersigwald replied:
Kind of! The semantics of "what does it mean to hook onto a project's entry-point target with AfterBuild" are very muddy.

Were you seeing execution return, or was it just continuing past an error in AfterTargets="Build"? If so, that's microsoft/msbuild#3345.

dougbu replied:
The occasional failures looked like a race condition between referenced projects building (including document generation) and the client project starting on its `BeforeTargets="BeforeCompile"` work (getting filenames of generated documents). To see the output I recall, the server project must have been running two targets simultaneously -- `GenerateOpenApiDocuments` and `OpenApiGetDocuments`. In particular, I saw a [`<Message/>`](https://github.com/aspnet/AspNetCore/pull/13364/files#diff-962355bc963c7ef291c534e32b08cb77R61) about tool execution just before or just after an [`<Error/>`](https://github.com/aspnet/AspNetCore/pull/13364/files#diff-962355bc963c7ef291c534e32b08cb77R28) about a missing file that the command writes.

Put another way, I was seeing the occasional "client" (project with an `@(OpenApiProjectReference)` item) build continue before the "server" (referenced project using this code) build completed. No errors until the client called the `GetOpenApiDocuments` target when the server had not finished running `GenerateOpenApiDocuments`.

rainersigwald replied:
Interesting. I think something deeper is wrong in this case. Any chance you have an MSBuild binlog of a build like this? Even a build that didn't fail due to the race would be instructive.

dougbu replied:
I'll get something to you early next week

dougbu reviewed (commented):
@rainersigwald please let us know if you have additional suggestions

rainersigwald commented on src/Tools/Extensions.ApiDescription.Server/src/buildMultiTargeting/Microsoft.Extensions.ApiDescription.Server.targets:
```diff
@@ -33,8 +32,9 @@
+  <!-- DispatchToInnerBuilds is last target outer Build depends on. -->
   <Target Name="_GenerateOpenApiDocuments"
-      AfterTargets="Build"
+      AfterTargets="DispatchToInnerBuilds"
```
Could you instead use `BeforeTargets="Build"`? That keeps from depending on the confusing Dispatch target.

dougbu replied:
My aim here is to run `GenerateOpenApiDocuments` in just one inner build (with a TFM the tool supports) regardless of the number of entries in `$(TargetFrameworks)`.

The inner `GenerateOpenApiDocuments` target has to run _after_ the project is built. (The tool invoked is an "outside man" that executes the application with a different-than-normal entry point.) Is there another option that would be less confusing?

I guess one option would be to change https://github.com/aspnet/AspNetCore/pull/13364/files#diff-b3e35eb5bf39d86aaddd2a229f8cef8cR30 to `Targets="Build;GenerateOpenApiDocuments"` and use `BeforeTargets="Build"` here. Would that run the specific inner build again when `DispatchToInnerBuilds` executes? Would it even work?

rainersigwald replied:
**tl;dr**: I think I'm trying to say "`BeforeTargets="Build"` is a drop-in substitute for almost all uses of `AfterTargets="Build"`, because `Build` does nothing itself so ordering around it shouldn't matter".

----

Naming targets is really hard. The convention is to make them logical from the dev's perpective, which makes them _super confusing_ from the build maintainer's perspective.

Specifically, targets with simple names run _after_ all of the steps that make them up. So for example, `Build` should be the last target that runs when you do `/t:Build` (or default target). That means that saying `BeforeTargets="Build"` is a good way to say "run this target last thing before saying 'Build' is done". Because of the [details of target ordering](https://docs.microsoft.com/en-us/visualstudio/msbuild/target-build-order?view=vs-2019#determine-the-target-build-order), that will (in the absence of other constraints) run _after_ all the things that `Build` `DependsOn`, giving you the behavior I think you want in this case.

dougbu replied:
OK, I'll update this too

rainersigwald commented on src/Tools/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -44,7 +43,7 @@
         Projects="@(OpenApiProjectReference)"
         RebaseOutputs="true"
-        RemoveProperties="Configuration;Platform;RuntimeIdentifier;TargetFramework;TargetFrameworks">
+        RemoveProperties="Configuration;Platform;RuntimeIdentifier;TargetFrameworks;%(OpenApiProjectReference.GlobalPropertiesToRemove)">
```
Do you recall why `TargetFrameworks` is in this list? It should be harmless, but I don't know of a case where it's passed as a global property. Do I need to update my mental map on that?

dougbu replied:
Thought I noticed `TargetFrameworks` was defined in inner builds. But, if that's just coming from the setting in the project file and won't be passed along to the referenced projects, I can remove it.

Given your comments above about `Configuration` and `Platform` (well, `%(ShouldUnsetParentConfigurationAndPlatform)`), should I _remove_ `Configuration;Platform;` here?

rainersigwald replied:
Yes, you should, great catch!

rainersigwald replied:
Also, no need to bother removing TargetFrameworks. It should be harmless, so don't change it just before release. But my understanding is that while it will be set during both inner and outer builds for a multitargeted project, it should only be as a normal property, not global.

dougbu replied:
OK, will make the first change

dougbu committed:
PR comments
- don't remove `$(Configuration)` or `$(Platform)` global properties
- move completely away from `AfterTargets="[Core]Build"` in these .targets and projects

dougbu commented:
Getting this in. Only failures in required checks were either flaky-but-still-running-in-main or completely unrelated to what's changing here.
